lib/DXIL/DxilOperations.cpp

```diff
@@ -3143,13 +2785,42 @@ llvm::Attribute::AttrKind OP::GetMemAccessAttr(OpCode opCode) {
 }
 
 bool OP::IsOverloadLegal(OpCode opCode, Type *pType) {
-  if (!pType)
+  if (static_cast<unsigned>(opCode) >=
+      static_cast<unsigned>(OpCode::NumOpCodes))
     return false;
-  if (opCode == OpCode::NumOpCodes)
+  if (!pType)
     return false;
-  unsigned TypeSlot = GetTypeSlot(pType);
-  return TypeSlot != UINT_MAX &&
-         m_OpCodeProps[(unsigned)opCode].bAllowOverload[TypeSlot];
+  auto &OpProps = m_OpCodeProps[static_cast<unsigned>(opCode)];
+
+  if (OpProps.NumOverloadDims == 0)
+    return pType->isVoidTy();
+
+  // Normalize 1+ overload dimensions into array.
+  Type *Types[DXIL::kDxilMaxOloadDims] = {pType};
+  if (OpProps.NumOverloadDims > 1) {
+    StructType *ST = dyn_cast<StructType>(pType);
+    // Make sure multi-overload is well-formed.
+    if (!ST || ST->hasName() ||
+        ST->getNumContainedTypes() != OpProps.NumOverloadDims)
+      return false;
+    for (unsigned I = 0; I < ST->getNumElements(); ++I)
+      Types[I] = ST->getElementType(I);
+  }
+
+  for (unsigned I = 0; I < OpProps.NumOverloadDims; ++I) {
+    Type *Ty = Types[I];
+    unsigned TypeSlot = GetTypeSlot(Ty);
+    if (!OpProps.AllowedOverloads[I][TypeSlot])
```

Comment:
I'm afraid I've run into a problem with this. The vector bit isn't set in the SlotMask for vector load/stores and I don't think it ever is. Either we need to check for the vector case first or change the generation so that bit is set.

Reply:
Fixed, thanks!